kobo/apps/project_ownership/models/transfer.py

```diff
@@ -0,0 +1,328 @@
+        on_delete=models.CASCADE,
+    )
+    invite = models.ForeignKey(
+        Invite,
+        related_name='transfers',
+        on_delete=models.CASCADE,
+    )
+
+    class Meta:
+        verbose_name = 'project ownership transfer'
+
+    def __str__(self) -> str:
+        return (
+            f'{self.asset}: '
+            f'{self.invite.sender.username} -> '
+            f'{self.invite.recipient.username}'
+        )
+
+    def process(self):
+        if self.status != TransferStatusChoices.PENDING.value:
+            raise TransferAlreadyProcessedException()
+
+        self.status = TransferStatusChoices.IN_PROGRESS.value
+        new_owner = self.invite.recipient
+        success = False
+        try:
+            if not self.asset.has_deployment:
+                with transaction.atomic():
+                    self._reassign_project_permissions(update_deployment=False)
+                    self._sent_app_in_messages()
+                    # FIXME - Draft stay in queue forever
+                    # Set async tasks to done for submissions and attachments.
+            else:
+                with transaction.atomic():
+                    with kc_transaction_atomic():
+                        deployment = self.asset.deployment
+                        with deployment.suspend_submissions(
+                            [self.asset.owner_id, new_owner.pk]
+                        ):
+                            # Update counters
+                            deployment.transfer_counters_ownership(new_owner)
+                            previous_owner_username = self.asset.owner.username
+                            self._reassign_project_permissions(
+                                update_deployment=True
+                            )
+                            deployment.rename_enketo_id_key(previous_owner_username)
+
+                        self._sent_app_in_messages()
+
+                # Run background tasks.
+                # 1) Rewrite `_userform_id` in MongoDB
+                async_task.delay(
+                    self.pk, TransferStatusTypeChoices.SUBMISSIONS.value
+                )
+
+            # 2) Move media files to new owner's home directory
+            async_task.delay(
+                self.pk, TransferStatusTypeChoices.MEDIA_FILES.value
+            )
+
+            success = True
+        finally:
+            if not success:
+                # We do not know which error has been raised, so no logs are
+                # saved. Sentry is our friend to find out what's going on.
+                self.status = (
+                    TransferStatusChoices.FAILED.value,
+                    'Error occurred while processing transfer',
+                )
+
+    def save(self, *args, **kwargs):
+
+        is_new = self.pk is None
+
+        super().save(*args, **kwargs)
+
+        if is_new:
+            self._init_statuses()
+
+    @property
+    def status(self):
+        if hasattr(self, 'prefetched_status'):
```

Comment:
I don't understand the complexity here, why does it sometimes have the ORM prefetched attribute?

Reply:
Moved it to the serializer!